### ISSUE TITLE
Preflighted requests support

### ISSUE THREAD

**Issue by YoukaiCat** [spenibus/cors-everywhere-firefox-addon#11]
Hello.
When I tried to set the progress event handlers, I got the following error:

`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at <URL> (Reason: CORS preflight channel did not succeed).  (unknown)`

My code:
```javascript
var request = new XMLHttpRequest();
request.responseType = "document";
request.upload.onprogress = function(event) { ... }
request.onprogress = function(event) { ... }
request.withCredentials = true;
request.open("POST", url);
request.send(formData);
```

According to the information that I found, browsers, at least Firefox and Chrome, perform "preflighted" requests using the OPTIONS method when progress event handler is set:

https://bugs.chromium.org/p/chromium/issues/detail?id=112585
https://bugzilla.mozilla.org/show_bug.cgi?id=727412

It would be nice if cors-everywhere support such requests by generating fake server answer with required Access-Control-Allow-* headers and 200-code.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Preflighted_requests
https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS

Thanks.

**Comment by spenibus** [comment]
Interesting. This was the topic of an email I received in January.

I tested this after receiving the email I mentioned earlier and it worked just fine at the time, when the resource returned a 2xx http code. The preflight request did however fail when I forced the server to return a 1xx, 3xx, 4xx or 5xx http code.

So my question as this point would be:
- Does the resource targeted by your preflight request return a 2xx http code ?

This addon, once activated, processes every http response that goes through the browser, in every tab, in every window. Which means that preflight requests are also treated. Although it does not alter http codes.

**Comment by YoukaiCat** [comment]
You are right: if the server returns a 200 HTTP code then everything works. 
Some servers do not support the OPTIONS method at all, while they support the GET and POST methods that are "really" needed to perform this request.

**Comment by spenibus** [comment]
Turns out modifying the http status code response is a lot less trivial than expected. We'll see how it goes. No time frame.

**Comment by spenibus** [comment]
Oh yeah, forgot about this.
Preliminary testing done a while back shows little to no possibility of modifying the response code.

I think I also tested altering the request to either avoid preflight or make preflight look fine and it didn't go well.

**Comment by titos1** [comment]
It would be really helpful if we could somehow suppress OPTIONS requests before every single CORS-Everywhere-d POST or GET request to a server that doesn't support the OPTIONS pre-flight request. Otherwise we're talking twice as many requests, every time.

Maybe setting a fake [Access-Control-Max-Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age) response header to the max allowed value of 86400 would silence it for that many seconds?
